private/bufpkg/bufcheck/buflint/internal/buflintvalidate/cel.go

```diff
@@ -0,0 +1,311 @@
+			cel.Types(dynamicpb.NewMessage(fieldDesc.Message())),
+			cel.Variable("this", cel.ObjectType(string(fieldDesc.Message().FullName()))),
+		)
+	} else {
+		env, err = env.Extend(
+			cel.Variable("this", protoKindToCELType(fieldDesc.Kind())),
+		)
+	}
+	if err != nil {
+		return err
+	}
+	for i, cel := range fieldConstraints.GetCel() {
+		celLocation := field.OptionExtensionLocation(
+			fieldConstraintsExtensionType,
+			celFieldTagInFieldConstraints,
+			int32(i),
+		)
+		if len(strings.TrimSpace(cel.Expression)) == 0 {
+			add(field, celLocation, nil, "cel expression is empty")
+			continue
+		}
+		ast, compileIssues := env.Compile(cel.Expression)
+		switch ast.OutputType() {
+		case types.BoolType, types.StringType, types.ErrorType:
+			// If type is types.ErrorType, compilation has failed and we will
+			// only add the compilation issues.
+		default:
+			add(field, celLocation, nil, "cel expression evaluates to unsupported type: %v", ast.OutputType())
+		}
+		if compileIssues.Err() != nil {
+			for _, parsedIssue := range parseCelIssuesText(compileIssues.Err().Error()) {
+				add(field, celLocation, nil, parsedIssue)
+			}
+		}
+	}
+	return nil
+}
+
+func getReflectMessageDescriptor(resolver protodesc.Resolver, message protosource.Message) (protoreflect.MessageDescriptor, error) {
+	descriptor, err := resolver.FindDescriptorByName(protoreflect.FullName(message.FullName()))
+	if err == protoregistry.NotFound {
+		return nil, fmt.Errorf("unable to resolve MessageDescriptor: %s", message.FullName())
+	}
+	if err != nil {
+		return nil, err
+	}
+	messageDescriptor, ok := descriptor.(protoreflect.MessageDescriptor)
+	if !ok {
+		// this should not happen
+		return nil, fmt.Errorf("%s is not a message", descriptor.FullName())
+	}
+	return messageDescriptor, nil
+}
+
+func getReflectFieldDescriptor(resolver protodesc.Resolver, field protosource.Field) (protoreflect.FieldDescriptor, error) {
+	descriptor, err := resolver.FindDescriptorByName(protoreflect.FullName(field.FullName()))
+	if err == protoregistry.NotFound {
+		return nil, fmt.Errorf("unable to resolve FieldDescriptor: %s", field.FullName())
+	}
+	if err != nil {
+		return nil, err
+	}
+	fieldDescriptor, ok := descriptor.(protoreflect.FieldDescriptor)
+	if !ok {
+		// this should never happen
+		return nil, fmt.Errorf("%s is not a field", descriptor.FullName())
+	}
+	return fieldDescriptor, nil
+}
+
+func getMessageConstraints(message protosource.Message) (*validate.MessageConstraints, bool, error) {
+	messageConstraintsMessageUntyped, found := message.OptionExtension(messageConstraintsExtensionType)
+	if !found {
+		return nil, false, nil
+	}
+	messageConstraintsMessage, ok := messageConstraintsMessageUntyped.(protoreflect.Message)
+	if !ok {
+		// this should never happen
+		return nil, false, fmt.Errorf("field extension expected to be `protoreflect.Message`, but has type %T", messageConstraintsMessageUntyped)
+	}
+	data, err := proto.Marshal(messageConstraintsMessage.Interface())
```

Comment:
This seems to do the same thing as in @rodaine's recent PR in protovalidate itself. Could you instead somehow use the protovalidate runtime to extract the constraints, instead of repeating all of this?

Reply:
You are right, I opened https://github.com/bufbuild/protovalidate-go/pull/59 and will use `DefaultResolver` to resolve these constraints.

Reply:
One implication of using `DefaultResolver` from `protovalidate-go` is that in `buf lint` we will need to use remote package for `protovalidate`, since `protovalidate` uses it.

Reply:
Resolved in [5e33b02](https://github.com/bufbuild/buf/pull/2473/commits/5e33b02a5044a09ee659acb1d7bde7395490ffe1)